src/neurostatslib/glm.py

```diff
@@ -1,431 +1,598 @@
-            raise ValueError(
-                "init_spikes has the wrong number of time steps!"
-                f"init_spikes time steps: {init_spikes.shape[1]}, "
-                f"spike_basis_matrix window size: {self.spike_basis_matrix.shape[1]}"
+
+        subkeys = jax.random.split(random_key, num=feedforward_input.shape[0])
+        # (n_samples, n_neurons)
+        feed_forward_contrib = jnp.einsum("ik,tik->ti", Wf, feedforward_input)
+
+        def scan_fn(
+            data: Tuple[jnp.ndarray, int], key: jax.random.PRNGKeyArray
+        ) -> Tuple[Tuple[jnp.ndarray, int], Tuple[jnp.ndarray, jnp.ndarray]]:
+            """Scan over time steps and simulate activity and rates.
+
+            This function simulates the neural activity and firing rates for each time step
+            based on the previous activity, feedforward input, and model coefficients.
+            """
+            activity, chunk = data
+
+            # Convolve the neural activity with the coupling basis matrix
+            conv_act = convolve_1d_trials(coupling_basis_matrix, activity[None, :, :])[
+                0
+            ]
+
+            # Extract the corresponding slice of the feedforward input for the current time step
+            input_slice = jax.lax.dynamic_slice(
+                feed_forward_contrib,
+                (chunk, 0),
+                (1, feed_forward_contrib.shape[1]),
             )
 
-        subkeys = jax.random.split(random_key, num=n_timesteps)
+            # Reshape the convolved activity and concatenate with the input slice to form the model input
+            conv_act = jnp.tile(
+                conv_act.reshape(conv_act.shape[0], -1), conv_act.shape[1]
+            ).reshape(conv_act.shape[0], conv_act.shape[1], -1)
 
-        def scan_fn(spikes, key):
-            # (n_neurons, n_basis_funcs, 1)
-            X = jnp.transpose(
-                convolve_1d_trials(self.spike_basis_matrix.T, spikes.T[None, :, :])[0],
-                (1, 2, 0),
+            # Predict the firing rate using the model coefficients
+            # Doesn't use predict because the non-linearity needs
+            # to be applied after we add the feed forward input
+            firing_rate = self.noise_model.inverse_link_function(
+                jnp.einsum("ik,tik->ti", Wr, conv_act) + input_slice + bs[None, :]
             )
-            fr = self._predict((Ws, bs), X).squeeze(-1)
-            new_spikes = jax.random.poisson(key, fr)
-            concat_spikes = jnp.column_stack((spikes[:, 1:], new_spikes))
-            return concat_spikes, new_spikes
 
-        _, simulated_spikes = jax.lax.scan(scan_fn, init_spikes, subkeys)
+            # Simulate activity based on the predicted firing rate
+            new_act = self.noise_model.emission_probability(key, firing_rate)
 
-        return simulated_spikes.T
+            # Prepare the spikes for the next iteration (keeping the most recent spikes)
```

Comment:
I'm a bit confused by this comment, how exactly are we preparing for next iteration? just stacking together?

what's `chunk+1` doing at the end there?

Reply:
It's cutting the first index and attach the newly generate sample at the end.  
chunk is the index which is not explicitly generated by  the `scan` and it is used to grab the feed-forward input for a specific sample.
I'll add a clarification

Reply:
I renamed chunk to t_sample